try to update to Elm 0.19

jschomay opened the pull request:
#20 
~~I've made the compiler happy, but the tests wont run - it gives a `maximum call stack` error, and I'm not sure why...~~

I made some judgement calls on some of the refactoring which may or may not be correct/desired.  I've added comments to this PR in those cases.

jschomay commented on src/Graph.elm:
```diff
@@ -243,19 +206,22 @@ computeEdgeDiff old new =
                         ( Just (Remove _), Remove _ ) ->
-                            Debug.crash "Graph.computeEdgeDiff: Collected two removals for the same edge. This is an error in the implementation of Graph and you should file a bug report!"
+                            -- Debug.crash "Graph.computeEdgeDiff: Collected two removals for the same edge. This is an error in the implementation of Graph and you should file a bug report!"
+                            Nothing
```
Not sure if Nothing is the best default (here and below).

sgraf812 replied:
The cases that would lead to a crash should return `Just eu`, where `eu` is the the value of `edgeUpdate label`. You could probably merge that with the last case and just have that case match on `(_, eu)` instead of `(Nothing, eu)`. This is strictly worse than what we had before, but at least it's compatible with Elm 0.19.

jschomay replied:
Should this go through the crashLoop too?

jschomay commented on src/Graph.elm:
```diff
@@ -684,10 +652,9 @@ mapContexts f =
 
-{-| Maps over node labels, possibly changing their types. Leaves the graph
-topology intact.
+{-| Maps over node labels Leaves the graph topology intact.
```
Note the change in API (0.19 won't allow changing record types).  Same applies for `mapEdges` below.

jschomay replied:
I managed to fix this

jschomay commented on src/Graph.elm:
```diff
@@ -730,39 +697,28 @@ type AcyclicGraph n e
 
 
-unsafeGet : String -> NodeId -> Graph n e -> NodeContext n e
-unsafeGet msg id graph =
```
This just doesn't work any more, so I made the calling code deal with it.

sgraf812 replied:
It's probably the best to instead call `crashLoop` (see below) with `msg`. Many of the crashes here have no real recovery case and it's better to loop endlessly than to return wrong results in the unlikely case that the invariants here are hurt.

jschomay commented on src/Graph.elm:
```diff
@@ -730,39 +697,28 @@ type AcyclicGraph n e
-
-                backEdges =
-                    IntDict.intersect ctx.outgoing backSetWithId
+                backEdge =
```
This is how I dealt with the removal of `unsafeGet`.  However, doing so ended up changing the error case of the returned `Result`.  I am confused why I couldn't construct the Edge in the error case below, but the compiler complained that it was a `Err (Edge ())` instead of a `Err (Edge e)`.  This seems connected to the `()` on line 706 above.  Since we don't know what type `e` is, we can't change that, so I had to change the type of the whole function, and sadly, lose some useful functionality.

sgraf812 replied:
Note that it's not `Err (Edge ())` but `Err ()`. That totally makes half the use cases of this function useless. The error you're seeing is probably related to the fact that you swapped the arguments to `IntDict.intersect`, which is left biased. We don't want to return the meaningless `()` values of the `backSetWithId`, but rather the values of the `ctx.outgoing` `IntDict`.

jschomay replied:
Good catch!  That works fine now.

jschomay replied:
Though if I use the crashLoop then this isn't necessary to change.  Is that your preference?

jschomay commented on src/Graph.elm:
```diff
@@ -1024,8 +979,9 @@ dfsTree seed graph =
-            Debug.crash "dfsTree: There can't be more than one DFS tree. This invariant is violated, please report this bug."
+        tree :: _ ->
+            -- Debug.crash "dfsTree: There can't be more than one DFS tree. This invariant is violated, please report this bug."
+            tree
```
This will quietly fail, but I don't think we can do any better.

sgraf812 replied:
Yeah, you're probably right. Although you could replace this by an inifinite loop instead (also applies to the other `Debug.crash` cases), probably makes more sense.

jschomay commented on src/Graph.elm:
```diff
@@ -1083,7 +1039,8 @@ ignorePath : SimpleNodeVisitor n e acc -> BfsNodeVisitor n e acc
     case path of
         [] ->
-            Debug.crash "Graph.ignorePath: No algorithm should ever pass an empty path into this BfsNodeVisitor."
+            -- Debug.crash "Graph.ignorePath: No algorithm should ever pass an empty path into this BfsNodeVisitor."
```
This will quietly fail.

sgraf812 replied:
Again, I'd probably prefer an infinite loop here.

sgraf812 replied:
Just create an auxiliary binding `crashLoop : String -> a; crashLoop reason = crashLoop reason`. It's crazy, but the only solution to this phony 'there are no crashes' madness.

jschomay commented on src/Graph.elm:
```diff
@@ -1217,12 +1175,13 @@ heightLevels (AcyclicGraph graph _) =
                         Nothing ->
-                            Debug.crash "Graph.heightLevels: Could not get a node of a graph which should be there by invariants. Please file a bug report!"
+                            -- Debug.crash "Graph.heightLevels: Could not get a node of a graph which should be there by invariants. Please file a bug report!"
+                            ( nextLevel, indegreesDec )
```
This will silently fail.

sgraf812 replied:
dito. Please call the `crashLoop` with appropriate `reason`, so that people can figure out why it crashed when looking at it in the debugger.

jschomay commented on src/Graph.elm:
```diff
@@ -1245,12 +1204,13 @@ heightLevels (AcyclicGraph graph _) =
                     case go currentLevel1 nextLevel1 indegrees1 (remove source.node.id graph) of
                         [] ->
-                            Debug.crash "Graph.heightLevels: Reached a branch which is impossible by invariants. Please file a bug report!"
+                            -- Debug.crash "Graph.heightLevels: Reached a branch which is impossible by invariants. Please file a bug report!"
```
This will silently fail

sgraf812 replied:
see above

jschomay commented on src/Graph.elm:
```diff
@@ -1263,7 +1223,7 @@ topologicalSort (AcyclicGraph graph ordering) =
             "Graph.topologicalSort: Invalid `AcyclicGraph`, where the ordering contained nodes not present in the graph"
     in
-    List.map (\id -> unsafeGet error id graph) ordering
+    List.filterMap (\id -> get id graph) ordering
```
This will silently filter out ids that don't exist

sgraf812 replied:
Keep the call to `unsafeGet` and make `unsafeGet` call `crashLoop`

jschomay commented:
I'm working on the tests, I've got some of the running (and I fixed the format mistake).  I"ll push a new commit up soon...

jhrcek commented on src/Graph.elm:
```diff
@@ -699,10 +666,10 @@ mapNodes f =
 topology intact.
 -}
-mapEdges : (e1 -> e2) -> Graph n e1 -> Graph n e2
+mapEdges : (e -> e) -> Graph n e -> Graph n e
```
Is there any reason to make the type more restrictive? With previous type signature people could change the type of edge labels. With this change it won't be possible any more. The same comment applies to mapNodes above.

sgraf812 replied:
I wonder this as well.

jschomay replied:
I agree that this is an unfortunate loss of flexibility.  The issue is that 0.19 no longer allows you to change the type of records in this way (https://github.com/elm/compiler/blob/master/upgrade-docs/0.19.md#stricter-record-update-syntax).

I figured out how to make it work, so I've restored the original type (for mapNodes too)

sgraf812 commented on tests/Tests/Graph/DOT.elm:
```diff
@@ -46,7 +46,7 @@ all =
   1 [label="To"]
   2 [label="Web"]
-  3 [label="\\"GraphViz\\"!"]
+  3 [label=""GraphViz"!"]
```
This is really a subtle breaking change I'm not willing to make. Try to properly fix https://github.com/elm-community/graph/pull/21/files#diff-4f0eb98e08fb690b622716455626060aL128.

jschomay replied:
I don't think that actually changes the behavior.  The change on L128 just wraps the string in quotes, which is what `Basics.toString` was doing since it was already a String value.

I think it is more of an issue with how escaping works in the test file.  I changed it to `"\\\"GraphViz\\\"!"` which will turn into the string `\"GraphViz\"!`, which is what I think the desired output is. I have tested this in https://dreampuf.github.io/GraphvizOnline/ and it works as I would expect.  

Let me know if I'm missing something.

jschomay commented:
I've made changes based on your feedback.  I managed to fix most of the issues.  The only thing I'm uncomfortable about is the crashLoop hack.  Is there any other way to approach that?
Let me know what you think of these changes.

jhrcek commented on src/Graph.elm:
```diff
@@ -648,7 +614,7 @@ so that the fold can exit early when the suspended accumulator is not forced.
 fold f acc graph =
     let
-        go acc graph1 =
+        go acc1 graph1 =
```
@sgraf812 Does this "closing over f" by creating this local helper `go` function have any observable effect? I mean in comparison with the same without that helper? Just read [this article](https://github.com/evancz/functional-programming-in-elm/blob/master/recursion/tail-call-elimination.md) and from what I understood the version below should work just the same. Just curious - I'll be happy if you prove me wrong and I'll learn something :smile:

```elm

fold : (NodeContext n e -> acc -> acc) -> acc -> Graph n e -> acc
fold f acc graph =
            let
                maybeContext =
                    graph
                        |> nodeIdRange
                        |> Maybe.map Tuple.first
                        |> Maybe.andThen (\id -> get id graph)
            in
            case maybeContext of
                Just ctx ->
                    fold f (f ctx acc) (remove ctx.node.id graph)

                Nothing ->
                    acc
```

sgraf812 replied:
It has no effect on whether the tail calls can be eliminated, if that is what you mean. I'm not too familiar with Elm's code gen, but I think it would amount to the same code. I mostly wrote it that way to safe characters (probably a bad excuse).

With an optimizing compiler like GHC, there is another perspective, though: The `fold` above is non-recursive and thus can be inlined, which in turn specialises `f` to the concrete argument at the call site. This in turn might enable more optimizations in the call `f ctx acc`, because now `f` is known (e.g. something calling + or whatever). I don't think Elm does anything in that regard.

jhrcek commented on src/Graph.elm:
```diff
@@ -1301,16 +1285,9 @@ stronglyConnectedComponents graph =
+toString nodeToString edgeToString graph =
     "Graph.fromNodesAndEdges "
-        ++ Basics.toString nodeList
+        ++ (String.join ", " <| List.map (\{ id, label } -> "Node " ++ String.fromInt id ++ " " ++ nodeToString label) <| nodes graph)
```
The beauty of the previous implementation was, that `toString graph` gave you the expression that you could have just copy-paste into elm file and it would be valid expression for creating the graph. This will no longer be true with this implementation. We could still try something like
`showNode  n = "Node { id = " ++ String.fromInt n.id ++ ...`
The only difficulty might be properly escaping characters in Strings. Is this something we should strive for?

jschomay replied:
Is this something you want?  I'm not sure what the use case would be.  Presumably one would have the code for the graph already, and not need to print it and copy it to get the literal code to make it.

Stepping back, I am wondering what`Graph.toString` is intended for, especially with `Graph.DOT.output`.   The former seems to be useful for debugging (note that you could just use `Debug.log`, though that isn't very useful), the latter seems useful for exporting to graphViz (I have personally done that a couple times in my projects).  However, could the graphViz version replace `Graph.toString`?  It seems odd to have two ways to print out the graph.

(related feature request would be to import from DOT)

jschomay replied:
Also if you want to preserve the ability to copy and paste the output, the clients will have to provide functions to turn their labels into the same copy/paste format, which dilutes the feature.

jschomay replied:
That said, I changed it to this format (which I think is useful for debugging:

    Graph [Node 0 (first), Node 1 (second), Node 2 (third)] [Edge 0->2 (relationship), Edge 0->1]

Note that labels are optional now.

jhrcek commented on src/Graph/DOT.elm:
```diff
@@ -125,13 +120,14 @@ outputWithStylesAndAttributes styles nodeAttrs edgeAttrs graph =
         attrAssocs =
             Dict.toList
-                >> List.map (\( k, v ) -> k ++ "=" ++ Basics.toString v)
+                >> List.map (\( k, v ) -> k ++ "=\"" ++ v ++ "\"")
```
I'm afraid this might break in some cases (e.g. when `v` String value itself contains '"' character, this will produce invalid dot). the `Basics.toString` was kind of hack to ensure proper escaping of special chars like that withing the String. Not sure how to replicate that behavior without generic toString now. I'm asking around on slack and looking for was to implement that and will get back soon..

jhrcek replied:
So people on elm-slack gave me this to escape Strings:
`Json.Encode.encode 0 (Json.Encode.string "I say \"hi\"!")`
Not sure it is worth the trouble to incur dependency on `elm\json` package though.

An alternative would be to use elm parser:
```elm
Parser.loop "" (\soFar ->
  Parser.oneOf
    [ Parser.end |> Parser.map (\_ -> Parser.Done soFar)
    , Parser.token "\\" |> Parser.map (\_ -> Parser.Loop (soFar ++ "\\\\"))
    , Parser.token "\"" |> Parser.map (\_ -> Parser.Loop (soFar ++ "\\\""))
    , Parser.chompWhile (c -> c /= '\\' && c /= '"')
        |> Parser.getChompedString
        |> Parser.map (\newStuff -> (Parser.Loop (soFar ++ newStuff))
    ])
```

jhrcek replied:
So I looked at the [DOT language specification](https://graphviz.gitlab.io/_pages/doc/info/lang.html) and it has this to say about escaping Strings:

> In quoted strings in DOT, the only escaped character is double-quote ("). That is, in quoted strings, the dyad \" is converted to "; all other characters are left unchanged. In particular, \\ remains \\. Layout engines may apply additional escape sequences.

So the solution would be trivial - just use something like
```elm
escapeDotString :: String -> String
escapeDotString = String.replace "\"" "\\\""
```
at all places where you're rendering Strings to dot file.

jschomay replied:
I started writing some tests for this, but ended up scrapping it for the following reasons:
- I literally cannot write the test without escaping things correctly (`"this has a \" and a \\ in it"`) or it fails to compile, and when the string is written like that, the tests would work as expected as is without the code to escape it
- I cannot write the expected cases (using multiline strings) with the escaping characters in place, because the formatter removes them, and if I use normal strings then I am just comparing two literals that are already the same
- The first two arguments to `Graph.DOT.output` are `n -> Maybe String` and `e -> Maybe String`, so the client should already be taking care of escaping strings in those functions, or ensuring they pass valid DOT strings in to begin with
- A similar thing applies to `outputWithStylesAndAttributes` - the client supplies the conversion to strings, we just need to add the DOT "glue" between those things, and the documentation say that.

(Note, when I tested this in https://dreampuf.github.io/GraphvizOnline/ I needed to escape backslashes as well as double quotes for it to compile.)

sgraf812 replied:
Seems like that site isn't really adhering to the DOT standard.

> I literally cannot write the test without escaping things correctly ("this has a \" and a \\ in it") or it fails to compile, and when the string is written like that, the tests would work as expected as is without the code to escape it

I don't think your code above would format the string `"a\"b"` correctly as `"\"a\\\"b\""`. You should at least escape double quotes (for example by using `String.replace`, as above). I'd say we take the slight incompatability with that GraphViz of not escaping back slashes, so it's really just `"=\"" ++ String.replace "\"" "\\\"" v ++ "\""`.

> I cannot write the expected cases (using multiline strings) with the escaping characters in place, because the formatter removes them, and if I use normal strings then I am just comparing two literals that are already the same

Sorry, I don't follow here. Isn't that exactly what the test suite was supposed to catch? An example would help.

> the client should already be taking care of escaping strings

I don't think that the client should be bothered with knowing how to escape strings in DOT files. That's what this library is for. You would expect the same from a JSON API.

> we just need to add the DOT "glue" between those things, and the documentation say that.

I think you're referring to this part of the documentation:

> Note that you have to take care of setting the appropriate node and edge labels
yourself.

I'm not the author of that module, but I suppose this means you have to set the `node` and `edge` labels, not that you have to escape the values they are set to.

Regardless, I'll just take it from here.

jschomay replied:
Ok, take a poke at it.  Just to clarify some points:

* If I gave a label to a node that looked  like `Node 0 "a\"b"` and ran it through the output code (as it is), and copy/paste the result into graphViz, the label on the rendered node would be `a"b` (the DOT string for that would be `label = "a\"b";`), which is what I would expect.  If I wanted the rendered label to be `a\"b` (and I don't know why I would want that) I would have to supply a label of `Node 0 "a\\\"b"` (for Elm to compile), which would turn out as expected, both in rendered graphViz and DOT.  
  In other words, since this is not valid Elm code: `Node 0 "a"b"`, I am forced to properly escape the strings for elm, which pass through to the DOT output.  *My main concern is that if we add a layer of escaping, if I made a `Node 0 "a\"b"` and the output was `label = "a\\\"b";` which renders as `a\"b` that would not be what I expect.*  So be sure to test that end to end (with https://dreampuf.github.io/GraphvizOnline/)

* After elm-format, `"""a\"b"""` becomes `"""a"b"""`, so I can't test that escaping is preserved if my expectations are in multiline strings.

I'm actually the original author of the DOT module (#8) though it has been modified since.

**Caveat: I have been looking at the output in the terminal when the tests run, and I think that has an extra layer of escaping done to it before printing it, so that may be having an effect.  When I have used this library, I have used the DOT in 3 ways - either by logging it in the console and copy/pasting, or rendering it into a textarea (or other element) to copy/paste, or passing it directly to graphViz (see https://github.com/jschomay/ene-graph/blob/master/src/graph.js).  I wouldn't be surprised if the escaping comes out differently between those 3 methods.  I haven't been able to test those 3 approaches now, but we definitely should.  I think having it work by passing it to the graphViz library is the most important behavior to have working (though for debugging I usually copy it out of the console).  Can you make an example app similar to the one I linked to to fully test this out before finalizing the approach?**

Let me know if you have any other questions.

sgraf812 replied:
> the label on the rendered node would be a"b

Are you sure about this? Here is a REPL session:

```Elm
> import Graph
> import Graph.DOT
> g = Graph.fromNodeLabelsAndEdgePairs [ "a\"b" ] []
Graph (Leaf { key = 0, value = { incoming = Empty, node = { id = 0, label = "a\"b" }, outgoing = Empty } })
    : Graph.Graph String ()
> Graph.DOT.output Just (always Nothing) g
"digraph G {\n  rankdir=TB\n  graph []\n  node []\n  edge []\n\n\n\n  0 [label=\"a\"b\"]\n}"
    : String
```

This is not what I would expect. Note that the REPL itself escapes the string here, so really the plain `String` returned is

```
digraph G {
  rankdir=TB
  graph []
  node []
  edge []



  0 [label="a"b"]
}
```

Paste this in the DOT formatter of your choice and be convinced this is not right. Granted, this means you can't just copy and paste from the REPL, but that's just a weakness of Elm's lacking console printing functionality (e.g. the difference in Haskell between calling `putStrLn` and `print`, which escapes string before printing them). If you would instead view the output string in a browser, you would get the exact same output I manually unescaped above.

> After elm-format, """a\"b""" becomes """a"b"""

Yuck, I remember having [hit this before](https://github.com/avh4/elm-format/issues/175). Seems like Elm treats \\" and " the same inside multi-line strings. The solution is to escape backslashes even in multi-line strings in order for them to appear, pretty much the way `Tests.Graph.DOT` currently does it. So, `"""a\\"b"""`.

> extra layer of escaping done to it before printing it

Ah yes, exactly. See above. I'm pretty sure that we should escape strings. It's a shame that the Elm REPL doesn't allow verbatim copying, but that's just due to lack of a proper printing API. What counts is the actual String that is generated, not how it's printed on the Terminal. If you put that String in a text area, it would show up as the manually unescaped string above. Therefore, we need to properly escape it in `output*`.

sgraf812 replied:
Btw., I realized that there's literally *no* `dot` implementation that only escapes " with \\. They all can't cope with nodes like (dot format) `0 [label="a\b"]` and need backslashes to be escaped, too (e.g. `0 [label="a\\b"]`). We should really find a library that can escape strings.

jhrcek commented on src/Graph.elm:
```diff
@@ -1301,16 +1285,9 @@ stronglyConnectedComponents graph =
-        edgeList =
-            edges graph
-    in
+toString : (n -> String) -> (e -> String) -> Graph n e -> String
```
It would be nice to have couple of unit tests for this more complex toString implementation.

jschomay replied:
What kind of unit tests do you have in mind?  The only difference is that you have to pass in your own converters.  I did use this for the test error messages, so even though there aren't specific tests for it, it is compiling.  I'll write a simple test to just prove it works correctly.

jschomay commented:
I added some more comments and one more commit around the `toString` function (because I changed the code around the comment, it got collapsed above.  See https://github.com/elm-community/graph/pull/21#discussion_r213039187 for the full discussion).

Let me know if you need anything else.

sgraf812 commented:
Thanks for your time and patience! I'll take it from here.

sgraf812 commented on tests/Tests.elm:
```diff
@@ -1,17 +0,0 @@
-module Tests exposing (suite)
-
-import Test exposing (..)
```
Ah, sorry, there's one more thing. What happened to this file? Does the new `elm-test` not need this anymore?

klazuka replied:
Correct. elm-test now uses some kind of magic to discover all modules which expose a value of type `Test` and then runs them. So this sort of file is no longer necessary.

jschomay replied:
No it doesn't.  It will run all tests exposed under `test/` (unless passed a specific file).  https://github.com/rtfeldman/node-test-runner

jschomay commented:
Yeah, I'm on board with that. The console print escaping is what was
throwing me off. Thanks for persisting.

On Tue, Aug 28, 2018, 1:07 PM Sebastian Graf <notifications@github.com>
wrote:

> *@sgraf812* commented on this pull request.
> ------------------------------
>
> In src/Graph/DOT.elm
> <https://github.com/elm-community/graph/pull/21#discussion_r213454243>:
>
> > @@ -125,13 +120,14 @@ outputWithStylesAndAttributes styles nodeAttrs edgeAttrs graph =
>          attrAssocs : Dict String String -> String
>          attrAssocs =
>              Dict.toList
> -                >> List.map (\( k, v ) -> k ++ "=" ++ Basics.toString v)
> +                >> List.map (\( k, v ) -> k ++ "=\"" ++ v ++ "\"")
>
> the label on the rendered node would be a"b
>
> Are you sure about this? Here is a REPL session:
>
> > import Graph> import Graph.DOT> g = Graph.fromNodeLabelsAndEdgePairs [ "a\"b" ] []Graph (Leaf { key = 0, value = { incoming = Empty, node = { id = 0, label = "a\"b" }, outgoing = Empty } })
>     : Graph.Graph String ()> Graph.DOT.output Just (always Nothing) g"digraph G {\n  rankdir=TB\n  graph []\n  node []\n  edge []\n\n\n\n  0 [label=\"a\"b\"]\n}"
>     : String
>
> This is not what I would expect. Note that the REPL itself escapes the
> string here, so really the plain String returned is
>
> digraph G {
>   rankdir=TB
>   graph []
>   node []
>   edge []
>
>
>
>   0 [label="a"b"]
> }
>
> Paste this in the DOT formatter of your choice and be convinced this is
> not right. Granted, this means you can't just copy and paste from the REPL,
> but that's just a weakness of Elm's lacking console printing functionality
> (e.g. the difference in Haskell between calling putStrLn and print, which
> escapes string before printing them). If you would instead view the output
> string in a browser, you would get the exact same output I manually
> unescaped above.
>
> After elm-format, """a"b""" becomes """a"b"""
>
> Yuck, I remember having hit this before
> <https://github.com/avh4/elm-format/issues/175>. Seems like Elm treats \"
> and " the same inside multi-line strings. The solution is to escape
> backslashes even in multi-line strings in order for them to appear, pretty
> much the way Tests.Graph.DOT currently does it. So, """a\\"b""".
>
> extra layer of escaping done to it before printing it
>
> Ah yes, exactly. See above. I'm pretty sure that we should escape strings.
> It's a shame that the Elm REPL doesn't allow verbatim copying, but that's
> just due to lack of a proper printing API. What counts is the actual String
> that is generated, not how it's printed on the Terminal. If you put that
> String in a text area, it would show up as the manually unescaped string
> above. Therefore, we need to properly escape it in output*.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elm-community/graph/pull/21#discussion_r213454243>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABva06nh_3TFdBjMfSbJpNpv5_TLKUbnks5uVaLqgaJpZM4WNJUK>
> .
>
